lsp/buy_channel.js

```diff
@@ -0,0 +1,210 @@
+          return cbk([400, 'ExpectedMessageToBuyChannel']);
+        }
+
+        if (!args.priority) {
+          return cbk([400, 'ExpectedPriorityToBuyChannel']);
+        }
+
+        if (!args.pubkey) {
+          return cbk([400, 'ExpectedPubkeyToBuyChannel']);
+        }
+
+        if (!args.tokens) {
+          return cbk([400, 'ExpectedTokensToBuyChannel']);
+        }
+
+        if (!args.type) {
+          return cbk([400, 'ExpectedTypeToBuyChannel']);
+        }
+
+        return cbk();
+      },
+
+      // Request info
+      validateMessage: ['validate', ({}, cbk) => {
+        const {message} = args;
+
+        if (!message.order_id) {
+          return cbk([400, 'ExpectedOrderIdToBuyChannel']);
+        }
+
+        if (Number(message.lsp_balance_sat) !== args.tokens) {
+          return cbk([400, 'ExpectedTokensToMatchLspBalance']);
+        }
+
+        if (message.client_balance_sat !== '' && message.client_balance_sat !== '0') {
+          return cbk([400, 'ExpectedZeroClientBalanceToBuyChannel']);
+        }
+
+        if (message.confirms_within_blocks !== args.priority) {
+          return cbk([400, 'ExpectedPriorityToMatchConfirmsWithinBlocks']);
+        }
+
+        if (message.channel_expiry_blocks !== constants.channelExpiryBlocks) {
+          return cbk([400, 'ExpectedMatchingChannelExpiryBlocksToBuyChannel']);
+        }
+
+        if (message.announce_channel !== args.announce_channel) {
+          return cbk([400, 'ExpectedMatchingAnnounceChannelToBuyChannel']);
+        }
+
+        if (message.order_state !== constants.orderStates.created) {
+          return cbk([400, 'ExpectedOrderStateToBeCreatedToBuyChannel']);
+        }
+
+        if (!message.payment) {
+          return cbk([400, 'ExpectedPaymentDetailsToBuyChannel']);
+        }
+
+        const {payment} = message;
+
+        // Fee total and order total sats should match because we don't support push amounts
+        if (payment.fee_total_sat !== payment.order_total_sat) {
+          return cbk([400, 'ExpectedMatchingFeeAndOrderTotalSatToBuyChannel']);
+        }
+
+        if (payment.state !== constants.paymentStates.expectPayment) {
+          return cbk([400, 'ExpectedExpectPaymentStateToBuyChannel']);
+        }
+
+        if (!payment.fee_total_sat) {
+          return cbk([400, 'ExpectedFeeTotalSatToBuyChannel']);
+        }
+
+        if (!payment.order_total_sat) {
+          return cbk([400, 'ExpectedOrderTotalSatToBuyChannel']);
+        }
+
+        try {
+          const res = parsePaymentRequest({request: payment.lightning_invoice});
+
+          if (res.tokens !== Number(payment.order_total_sat)) {
+            return cbk([400, 'ExpectedMatchingTokensInPaymentRequest']);
+          }
+        } catch (err) {
+          return cbk([400, 'ExpectedValidPaymentRequestToBuyChannel', {err}]);
+        }
+
+        return cbk();
+      }],
+
+      // Ask to buy the channel
+      ask: ['validateMessage', ({}, cbk) => {
```

Comment:
should probe first to see what the routing fee is going to cost and then feed that routing fee back into the pay call